### DISCUSSION TITLE
Ability to render objects on a Cell

### DISCUSSION THREAD

**Discussion by ruslanguns** [KevinVandy/mantine-react-table#4] [Show and tell]
Recently, I had the opportunity to work on a project utilizing Mantine-React-Table. One of the features I was able to add was an expandable cell that renders its content in a tooltip, similar to tables found in popular data visualization tools such as Kibana and Grafana.

One of the key factors that made this implementation possible was the use of the polymorphism approach in Mantine-React-Table. By utilizing this approach, I was able to replicate the expandable cell feature without much effort. This approach allows developers to easily customize and extend the functionality of the library to fit their specific needs.

![image](https://user-images.githubusercontent.com/19806968/213589078-4720ccc9-5491-4b0c-bd7e-f816f649cdd1.png)

**Comment by dbnar2** [comment]
Couldnt this be achievable by just calling a onRowClick and render a modal?

**Reply by ruslanguns** [reply]
If you have a "onRowClick" it will be triggering every time the user clicks on the whole row, indifferently of the cell, so in my case I solved by using the `accessorFn`: 

```ts
{
  header: 'Gateway Metadata',
  enableClickToCopy: false,

  accessorFn: (gatewayMetadata) => {
    return (
      <ExpandableCell
        title="Gateway Metadata"
        cellContent={() => <span>{JSON.stringify(gatewayMetadata)}</span>}
      >
        <JsonViewer data={gatewayMetadata} isDark={isDark} />
      </ExpandableCell>
    );
  },
},
```

This approach has few limitations though, for ie. if you want to enable the click to copy functionality it will be triggering the copy to clipboard even clicking in the ZoomIn Icon, which is not appropriate, this is because the click to copy renders all the content populated in the cell. So I had to disable it.

**Reply by KevinVandy** [reply]
I do not recommend putting JSX in the accessorFn. Instead, use the `Cell` column option to render - custom JSX. The accessorFn should only parse primitive values if you still want sorting and filtering to work.

**Reply by KevinVandy** [reply]
The click to copy stuff will start working again if you use Cell instead of accessorFn here too

**Reply by ruslanguns** [reply]
Sorry for late response. Ok understood, will remove it, and update it accordingly.

**Reply by ruslanguns** [reply]
I made the adjustments and now seems more organic, thanks for the tip.
![image](https://user-images.githubusercontent.com/19806968/215320281-1083684d-4495-4488-a757-1c18860832db.png)

Although noticed of a couple issues (after upgrading from 0.2.0 to 0.6.0):
![image](https://user-images.githubusercontent.com/19806968/215319751-d981c13c-3947-4436-86e3-59db840c0bad.png)

1. the popover position seems to be relative to the Table (I will look for a transport mechanism to render this outside the table) 
2. Before on 0.2.0 when cell content exceeded the width, a truncate with three dots was displayed.

**Reply by ruslanguns** [reply]
for instance I share the ExpandableCell just in case there's any clue of why this is happening:

```tsx
import { Popover, Text, Title } from '@mantine/core';
import { ZoomInIcon } from '@radix-ui/react-icons';
import { PropsWithChildren } from 'react';

type Props = {
  title: string;
  cell: unknown;
};

export default function ExpandableCell({
  title,
  cell,
  children,
}: Props & PropsWithChildren) {
  return (
    <div className="flex items-center">
      <Popover withArrow trapFocus shadow="md" position="bottom">
        <Popover.Target>
          <ZoomInIcon className="mr-2 cursor-pointer" />
        </Popover.Target>
        <Popover.Dropdown>
          <Title order={5} className="mb-2">
            {title}
          </Title>
          {children}
        </Popover.Dropdown>
      </Popover>
      <Text>{typeof cell === 'string' ? cell : JSON.stringify(cell)}</Text>
    </div>
  );
}
```

**Reply by KevinVandy** [reply]
I think your problem might be related to the `layoutMode` prop that I changed the default value of recently. Try `layoutMode="semantic"` (instead of the default "grid").

**Reply by ruslanguns** [reply]
Cool, that fixed my issue and now I'm using the latest version!

**Reply by ruslanguns** [reply]
Also the missing truncated dots were because I had the click to copy functionality enabled, by disabling it, it now shows the proper truncation.

**Reply by KevinVandy** [reply]
The layout mode was changed because of some Firefox row border issues. I kinda want to find a root solution to that instead of switching the who CSS layout to flexbox like now.

**Reply by ruslanguns** [reply]
That seems quite challenging
